src/um/icmpapi.rs

```diff
@@ -0,0 +1,77 @@
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// All files in the project carrying such notice may not be copied, modified, or distributed
+// except according to those terms.
+use shared::minwindef::{BOOL, DWORD, FARPROC, LPVOID, WORD};
+use shared::ntdef::{HANDLE, PVOID};
+use shared::ws2ipdef::PSOCKADDR_IN6_LH;
+use um::ipexport::IPAddr;
+#[cfg(target_pointer_width = "32")]
+use um::ipexport::PIP_OPTION_INFORMATION;
+#[cfg(target_pointer_width = "64")]
+use um::ipexport::PIP_OPTION_INFORMATION32 as PIP_OPTION_INFORMATION;
+extern "system" {
+    pub fn IcmpCreateFile() -> HANDLE;
+    pub fn Icmp6CreateFile() -> HANDLE;
+    pub fn IcmpCloseHandle(IcmpHandle: HANDLE) -> BOOL;
+    pub fn IcmpSendEcho(
+        IcmpHandle: HANDLE,
+        DestinationAddress: IPAddr,
+        RequestData: LPVOID,
+        RequestSize: WORD,
+        RequestOptions: PIP_OPTION_INFORMATION,
+        ReplyBuffer: LPVOID,
+        ReplySize: DWORD,
+        Timeout: DWORD
+    ) -> DWORD;
+    pub fn IcmpSendEcho2(
+        IcmpHandle: HANDLE,
+        Event: HANDLE,
+        ApcRoutine: FARPROC, // or PIO_APC_ROUTINE
+        ApcContext: PVOID,
+        DestinationAddress: IPAddr,
+        RequestData: LPVOID,
+        RequestSize: WORD,
+        RequestOptions: PIP_OPTION_INFORMATION,
+        ReplyBuffer: LPVOID,
+        ReplySize: DWORD,
+        Timeout: DWORD
+    ) -> DWORD;
+    pub fn IcmpSendEcho2Ex(
+        IcmpHandle: HANDLE,
+        Event: HANDLE,
+        ApcRoutine: FARPROC, // or PIO_APC_ROUTINE
```

Comment:
Meant to point this out originally... doing it now while I'm thinking about it.
The type of ApcRoutine should be PIO_APC_ROUTINE instead of FARPROC (for any OS after XP), however this type doesn't exist, and implementing it seems to be problematic. The (C) definition is ```typedef
VOID WINAPI
(*PIO_APC_ROUTINE) (
    IN PVOID ApcContext,
    IN PIO_STATUS_BLOCK IoStatusBlock,
    IN ULONG Reserved
    );``` but as far as I can see (P)IO_STATUS_BLOCK is not a type that is defined in this API, but instead comes from a kernel/driver API.

Because of this I left the type as FARPROC, although I could see changing it to PVOID or defining some dummy PIO_APC_ROUTINE.